nighthawk/source/common/sequencer_impl.cc

```diff
@@ -0,0 +1,151 @@
+              targets_initiated_, targets_completed_, rate);
+
+  } else {
+    ENVOY_LOG(
+        trace, "Processed {} operations in {} ms. ({} per second)", targets_completed_,
+        std::chrono::duration_cast<std::chrono::milliseconds>(time_source_.monotonicTime() - start_)
+            .count(),
+        rate);
+  }
+}
+
+void SequencerImpl::updateStatisticOnUnblockIfNeeded(const Envoy::MonotonicTime& now) {
+  if (blocked_) {
+    blocked_ = false;
+    blockedStatistic_.addValue((now - blocked_start_).count());
+  }
+}
+
+void SequencerImpl::updateStartBlockingTimeIfNeeded() {
+  if (!blocked_) {
+    blocked_ = true;
+    blocked_start_ = time_source_.monotonicTime();
+  }
+}
+
+void SequencerImpl::run(bool from_periodic_timer) {
+  ASSERT(running_);
+  const auto now = time_source_.monotonicTime();
+  const auto runtime = now - start_;
+
+  // If we exceed the benchmark duration.
+  if (runtime > duration_) {
+    if (targets_completed_ == targets_initiated_) {
+      // All work has completed. Stop this sequencer.
+      stop(false);
+    } else {
+      // After the benchmark duration has exceeded, we wait for a grace period for outstanding work
+      // to wrap up. If that takes too long we warn about it and quit.
+      if (runtime - duration_ > grace_timeout_) {
+        stop(true);
+        return;
+      }
+      if (from_periodic_timer) {
+        scheduleRun();
+      }
+    }
+    return;
+  }
+
+  while (rate_limiter_.tryAcquireOne()) {
+    // The rate limiter says it's OK to proceed and call the target. Let's see if the target is OK
+    // with that as well.
+    const bool target_could_start = target_([this, now]() {
+      const auto dur = time_source_.monotonicTime() - now;
+      latencyStatistic_.addValue(dur.count());
+      targets_completed_++;
+      // Immediately schedule us to check again, as chances are we can get on with the next task.
+      adhoc_timer_->enableTimer(0ms);
+    });
+
+    if (target_could_start) {
+      updateStatisticOnUnblockIfNeeded(now);
+      targets_initiated_++;
+    } else {
+      // This should only happen when we are running in closed-loop mode, which is always at the
+      // time of writing this. The target wasn't able to proceed. Update the rate limiter, we'll try
+      // again later.
+      updateStartBlockingTimeIfNeeded();
+      rate_limiter_.releaseOne();
+      break;
+    }
+  }
+
+  if (!from_periodic_timer) {
+    if (targets_initiated_ == targets_completed_) {
+      // We saturated the rate limiter, and there's no outstanding work.
+      // That means it looks like we are idle. Spin this event to improve
+      // accuracy. As a side-effect, this may help prevent CPU frequency scaling
+      // due to c-state changes. But on the other hand it may cause thermal throttling.
+      // TODO(oschaaf): Ideally we would have much finer grained timers instead.
+      // TODO(oschaaf): Optionize performing this spin loop.
+      platform_util_.yieldCurrentThread();
```

Comment:
Maybe that not yielding would be another nice option to have? It *does* behave a little annoying in the current tests, because we don't have a place to move simulated time forward when we drop the call, but that is solvable. 

On a machine tuned for running NH cores could be isolated for it to run on, in which case yielding is practically guaranteed to be a no-op, as there should be no other thread to yield to. [1]

Through some quick manual testing, I could not find an obvious difference in measured timings.I think this probably does make a difference in some setups/scenarios though (noisy systems with lots of things going on? single core/cpu?).

[1] http://man7.org/linux/man-pages/man2/sched_yield.2.html
```
       If the calling thread is the only thread in the highest priority list
       at that time, it will continue to run after a call to sched_yield().
```

Reply:
I think this is fine, I was just curious, nothing actionable beyond what you wrote above, thanks.